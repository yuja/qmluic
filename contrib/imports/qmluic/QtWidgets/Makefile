JQ = jq
QMAKE = qmake

QT_INSTALL_BINS = $(shell $(QMAKE) -query QT_INSTALL_BINS)
QT_INSTALL_LIBS = $(shell $(QMAKE) -query QT_INSTALL_LIBS)
QT_VERSION = $(shell $(QMAKE) -query QT_VERSION)

QMLTYPEREGISTRAR = $(QT_INSTALL_BINS)/qmltyperegistrar
QT_METATYPES = $(QT_INSTALL_LIBS)/metatypes

major_version = $(word 1,$(subst ., ,$(QT_VERSION)))
qtx = qt$(major_version)

.PHONY: all
all: plugins.qmltypes

plugins.qmltypes: metatypes.json $(QT_METATYPES)/$(qtx)core_metatypes.json
	$(QMLTYPEREGISTRAR) \
		--import-name qmluic.QtWidgets \
		--major-version $(major_version) \
		--foreign-types $(QT_METATYPES)/$(qtx)core_metatypes.json \
		--generate-qmltypes $@ $< > /dev/null

metatypes.json: $(QT_METATYPES)/$(qtx)widgets_metatypes.json
	$(JQ) 'map(.classes |= map({"classInfos": [{"name": "QML.Element", "value": "auto"}]} + .))' $< > $@

.PHONY: clean
clean:
	$(RM) plugins.qmltypes metatypes.json
