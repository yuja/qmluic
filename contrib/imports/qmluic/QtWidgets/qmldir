module qmluic.QtWidgets
typeinfo plugins.qmltypes
depends QtQml
