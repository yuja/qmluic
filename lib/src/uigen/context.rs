use super::objcode::ObjectCodeMap;
use crate::objtree::{ObjectNode, ObjectTree};
use crate::qmldoc::UiDocument;
use crate::qtname::FileNameRules;
use crate::typedexpr::{BuiltinFunctionKind, RefKind, RefSpace};
use crate::typemap::{
    Class, Enum, ImportedModuleSpace, ModuleId, NamedType, TypeMap, TypeMapError, TypeSpace,
};
use thiserror::Error;

/// How to process dynamic bindings.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum DynamicBindingHandling {
    /// Silently drops dynamic bindings.
    Omit,
    /// Generates C++ code to set up dynamic bindings.
    Generate,
    /// Generates errors for dynamic bindings.
    Reject,
}

/// Resources passed around the UI object constructors.
#[derive(Clone, Debug)]
pub struct BuildContext<'a> {
    pub(super) type_map: &'a TypeMap,
    pub file_name_rules: FileNameRules,
    pub dynamic_binding_handling: DynamicBindingHandling,
    pub(super) classes: KnownClasses<'a>,
}

#[derive(Clone, Debug)]
pub(super) struct BuildDocContext<'a, 't, 's> {
    pub type_name: &'s str,
    pub source: &'s str,
    pub file_name_rules: &'s FileNameRules,
    pub classes: &'s KnownClasses<'a>,
    pub type_space: &'s ImportedModuleSpace<'a>,
    pub object_tree: &'s ObjectTree<'a, 't>,
    object_code_maps: &'s [ObjectCodeMap<'a, 't, 's>],
}

/// Context where expression is supposed to be evaluated.
#[derive(Clone, Debug)]
pub(super) struct ObjectContext<'a, 't, 's> {
    pub source: &'s str,
    pub classes: &'s KnownClasses<'a>,
    pub type_space: &'s ImportedModuleSpace<'a>,
    pub object_tree: &'s ObjectTree<'a, 't>,
    obj_node: ObjectNode<'a, 't, 's>,
}

/// Classes to be used to switch uigen paths.
#[derive(Clone, Debug)]
pub(super) struct KnownClasses<'a> {
    pub action: Class<'a>,
    pub action_separator: Class<'a>,
    pub brush: Class<'a>,
    pub color: Class<'a>,
    pub combo_box: Class<'a>,
    pub cursor: Class<'a>,
    pub cursor_shape: Enum<'a>,
    pub font: Class<'a>,
    pub form_layout: Class<'a>,
    pub grid_layout: Class<'a>,
    pub hbox_layout: Class<'a>,
    pub icon: Class<'a>,
    pub key_sequence: Class<'a>,
    pub key_sequence_standard_key: Enum<'a>,
    pub layout: Class<'a>,
    pub list_widget: Class<'a>,
    pub margins: Class<'a>,
    pub menu: Class<'a>,
    pub palette: Class<'a>,
    pub pixmap: Class<'a>,
    pub push_button: Class<'a>,
    pub rect: Class<'a>,
    pub size: Class<'a>,
    pub size_policy: Class<'a>,
    pub spacer_item: Class<'a>,
    pub tab_widget: Class<'a>,
    pub table_view: Class<'a>,
    pub tree_view: Class<'a>,
    pub vbox_layout: Class<'a>,
    pub widget: Class<'a>,
}

impl<'a> BuildContext<'a> {
    /// Sets up context with the given `type_map`.
    pub fn prepare(
        type_map: &'a TypeMap,
        file_name_rules: FileNameRules,
        dynamic_binding_handling: DynamicBindingHandling,
    ) -> Result<Self, BuildContextError> {
        const MODULE_NAME: &str = "qmluic.QtWidgets";
        let module = type_map
            .get_module(&ModuleId::Named(MODULE_NAME.into()))
            .ok_or(BuildContextError::ModuleNotFound(MODULE_NAME))?;
        let get_class = |name| match module.get_type(name) {
            Some(Ok(NamedType::Class(cls))) => Ok(cls),
            Some(Err(e)) => Err(BuildContextError::TypeResolution(name, e)),
            Some(Ok(_)) | None => Err(BuildContextError::ClassNotFound(name)),
        };
        let get_enum = |scoped_name| {
            match module.get_type_scoped(scoped_name) {
                Some(Ok(NamedType::Enum(en))) => Ok(en),
                Some(Err(e)) => Err(BuildContextError::TypeResolution(scoped_name, e)),
                Some(Ok(_)) | None => {
                    // not a "class", but who cares
                    Err(BuildContextError::ClassNotFound(scoped_name))
                }
            }
        };
        let classes = KnownClasses {
            action: get_class("QAction")?,
            action_separator: get_class("QActionSeparator")?,
            brush: get_class("QBrush")?,
            color: get_class("QColor")?,
            combo_box: get_class("QComboBox")?,
            cursor: get_class("QCursor")?,
            cursor_shape: get_enum("Qt::CursorShape")?,
            font: get_class("QFont")?,
            form_layout: get_class("QFormLayout")?,
            grid_layout: get_class("QGridLayout")?,
            hbox_layout: get_class("QHBoxLayout")?,
            icon: get_class("QIcon")?,
            key_sequence: get_class("QKeySequence")?,
            key_sequence_standard_key: get_enum("QKeySequence::StandardKey")?,
            layout: get_class("QLayout")?,
            list_widget: get_class("QListWidget")?,
            margins: get_class("QMargins")?,
            menu: get_class("QMenu")?,
            palette: get_class("QPalette")?,
            pixmap: get_class("QPixmap")?,
            push_button: get_class("QPushButton")?,
            rect: get_class("QRect")?,
            size: get_class("QSize")?,
            size_policy: get_class("QSizePolicy")?,
            spacer_item: get_class("QSpacerItem")?,
            tab_widget: get_class("QTabWidget")?,
            table_view: get_class("QTableView")?,
            tree_view: get_class("QTreeView")?,
            vbox_layout: get_class("QVBoxLayout")?,
            widget: get_class("QWidget")?,
        };
        Ok(BuildContext {
            type_map,
            file_name_rules,
            dynamic_binding_handling,
            classes,
        })
    }
}

impl<'a, 't, 's> BuildDocContext<'a, 't, 's> {
    pub fn new(
        doc: &'s UiDocument,
        type_space: &'s ImportedModuleSpace<'a>,
        object_tree: &'s ObjectTree<'a, 't>,
        object_code_maps: &'s [ObjectCodeMap<'a, 't, 's>],
        base_ctx: &'s BuildContext<'a>,
    ) -> Self {
        BuildDocContext {
            type_name: doc.type_name(),
            source: doc.source(),
            file_name_rules: &base_ctx.file_name_rules,
            classes: &base_ctx.classes,
            type_space,
            object_tree,
            object_code_maps,
        }
    }

    pub fn code_map_for_object(&self, obj_node: ObjectNode) -> &ObjectCodeMap<'a, 't, 's> {
        &self.object_code_maps[obj_node.flat_index()]
    }

    pub fn make_object_context(
        &self,
        obj_node: ObjectNode<'a, 't, 's>,
    ) -> ObjectContext<'a, 't, 's> {
        ObjectContext {
            source: self.source,
            classes: self.classes,
            type_space: self.type_space,
            object_tree: self.object_tree,
            obj_node,
        }
    }
}

impl<'a, 't, 's> ObjectContext<'a, 't, 's> {
    pub fn new(
        doc: &'s UiDocument,
        type_space: &'s ImportedModuleSpace<'a>,
        object_tree: &'s ObjectTree<'a, 't>,
        obj_node: ObjectNode<'a, 't, 's>,
        base_ctx: &'s BuildContext<'a>,
    ) -> Self {
        ObjectContext {
            source: doc.source(),
            classes: &base_ctx.classes,
            type_space,
            object_tree,
            obj_node,
        }
    }
}

impl<'a> RefSpace<'a> for ObjectContext<'a, '_, '_> {
    fn get_ref(&self, name: &str) -> Option<Result<RefKind<'a>, TypeMapError>> {
        // object id lookup is explicit, which should precede any other implicit this lookups.
        let me = &self.obj_node;
        if let Some(obj) = self.object_tree.get_by_id(name) {
            Some(Ok(RefKind::Object(obj.class().clone())))
        } else if let Some(r) = me.class().get_property(name) {
            Some(r.map(|p| RefKind::ObjectProperty(me.class().clone(), me.name().to_owned(), p)))
        } else if let Some(r) = me.class().get_public_method(name) {
            Some(r.map(|m| RefKind::ObjectMethod(me.class().clone(), me.name().to_owned(), m)))
        } else if let Some(r) = self.type_space.get_type(name) {
            Some(r.map(RefKind::Type))
        } else if name == "qsTr" {
            Some(Ok(RefKind::BuiltinFunction(BuiltinFunctionKind::Tr)))
        } else {
            None
        }
    }
}

/// Error occurred while setting up [`BuildContext`].
#[derive(Clone, Debug, Error)]
pub enum BuildContextError {
    #[error("required class not found: {0}")]
    ClassNotFound(&'static str),
    #[error("required module not found: {0}")]
    ModuleNotFound(&'static str),
    #[error("required type '{0}' not resolved: {1}")]
    TypeResolution(&'static str, #[source] TypeMapError),
}
