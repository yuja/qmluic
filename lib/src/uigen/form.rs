use super::object::UiObject;
use super::xmlutil;
use super::{FileNameRules, XmlResult, XmlWriter};
use crate::typemap::{Class, TypeSpace};
use quick_xml::events::{BytesStart, Event};
use std::io;

/// Top-level object wrapper to be serialized to UI XML.
#[derive(Clone, Debug)]
pub struct UiForm {
    pub class: Option<String>,
    pub root_object: UiObject,
    pub custom_widgets: Vec<CustomWidget>,
}

impl UiForm {
    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let tag = BytesStart::borrowed_name(b"ui").with_attributes([("version", "4.0")]);
        writer.write_event(Event::Start(tag.to_borrowed()))?;
        if let Some(name) = &self.class {
            xmlutil::write_tagged_str(writer, "class", name)?;
        }
        self.root_object.serialize_to_xml(writer)?;
        if !self.custom_widgets.is_empty() {
            let tag = BytesStart::borrowed_name(b"customwidgets");
            writer.write_event(Event::Start(tag.to_borrowed()))?;
            for w in &self.custom_widgets {
                w.serialize_to_xml(writer)?;
            }
            writer.write_event(Event::End(tag.to_end()))?;
        }
        writer.write_event(Event::End(tag.to_end()))?;
        writer.write(b"\n")
    }
}

/// User type referenced from the [`UiForm`].
#[derive(Clone, Debug)]
pub struct CustomWidget {
    pub class: String,
    pub extends: String,
    pub header: String,
}

impl CustomWidget {
    pub(super) fn from_class(cls: &Class, file_name_rules: &FileNameRules) -> Option<Self> {
        // If super class doesn't exist, diagnostic message would have been emitted while
        // building the object representation. So returns silently.
        let super_cls = cls.public_super_classes().next()?;
        Some(CustomWidget {
            class: cls.qualified_cxx_name().into(),
            extends: super_cls.qualified_cxx_name().into(),
            header: file_name_rules.type_name_to_cxx_header_name(cls.name()),
        })
    }

    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let tag = BytesStart::borrowed_name(b"customwidget");
        writer.write_event(Event::Start(tag.to_borrowed()))?;
        xmlutil::write_tagged_str(writer, "class", &self.class)?;
        xmlutil::write_tagged_str(writer, "extends", &self.extends)?;
        xmlutil::write_tagged_str(writer, "header", &self.header)?;
        writer.write_event(Event::End(tag.to_end()))
    }
}
