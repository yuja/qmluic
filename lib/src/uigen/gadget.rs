use super::expr::{self, ConstantValue};
use super::xmlutil;
use super::{XmlResult, XmlWriter};
use crate::diagnostic::{Diagnostic, Diagnostics};
use crate::qmlast::{Node, UiBindingMap, UiBindingValue};
use crate::typemap::{Class, TypeSpace};
use quick_xml::events::{BytesStart, Event};
use std::collections::HashMap;
use std::io;

/// Constant map-like object which can be serialized to UI XML.
#[derive(Clone, Debug)]
pub struct ConstantGadget {
    pub name: String,
    pub properties: HashMap<String, ConstantValue>,
}

impl ConstantGadget {
    fn new(name: &str, properties: HashMap<String, ConstantValue>) -> Self {
        ConstantGadget {
            name: name.to_owned(),
            properties,
        }
    }

    /// Generates gadget of `cls` type from the given `binding_map`.
    pub fn from_binding_map(
        cls: &Class,
        node: Node,
        binding_map: &UiBindingMap,
        source: &str,
        diagnostics: &mut Diagnostics,
    ) -> Option<Self> {
        match cls.name() {
            "QRect" => Some(Self::new(
                "rect",
                collect_constant_properties(cls, binding_map, source, diagnostics),
            )),
            "QSize" => Some(Self::new(
                "size",
                collect_constant_properties(cls, binding_map, source, diagnostics),
            )),
            _ => {
                diagnostics.push(Diagnostic::error(
                    node.byte_range(),
                    format!("unsupported gadget type: {}", cls.qualified_name()),
                ));
                None
            }
        }
    }

    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let tag = BytesStart::borrowed_name(self.name.as_ref());
        writer.write_event(Event::Start(tag.to_borrowed()))?;
        let mut pairs: Vec<_> = self.properties.iter().collect();
        pairs.sort_by_key(|&(k, _)| k);
        for (k, v) in pairs {
            xmlutil::write_tagged_str(writer, k, v.to_string())?;
        }
        writer.write_event(Event::End(tag.to_end()))?;
        Ok(())
    }
}

/// Parses the given `binding_map` into a map of constant values.
///
/// Unparsable properties are excluded from the resulting map so as many diagnostic messages
/// will be generated as possible.
fn collect_constant_properties(
    cls: &Class,
    binding_map: &UiBindingMap,
    source: &str,
    diagnostics: &mut Diagnostics,
) -> HashMap<String, ConstantValue> {
    binding_map
        .iter()
        .filter_map(|(&name, value)| {
            if let Some(ty) = cls.get_property_type(name) {
                match value {
                    UiBindingValue::Node(n) => {
                        ConstantValue::from_expression(cls, &ty, *n, source, diagnostics)
                    }
                    UiBindingValue::Map(n, _) => {
                        diagnostics.push(Diagnostic::error(
                            n.byte_range(),
                            "binding map cannot be gadget property",
                        ));
                        None
                    }
                }
            } else {
                diagnostics.push(Diagnostic::error(
                    value.node().byte_range(),
                    format!(
                        "unknown property of class '{}': {}",
                        cls.qualified_name(),
                        name
                    ),
                ));
                None
            }
            .map(|v| (name.to_owned(), v))
        })
        .collect()
}

/// Constant size policy which can be serialized to UI XML.
///
/// `QSizePolicy` is special in that
/// - it is registered as gadget type but have no useful property,
/// - UI XML attribute names are diverged from the method/property names.
#[derive(Clone, Debug, Default)]
pub struct ConstantSizePolicy {
    pub horizontal_policy: Option<String>,
    pub vertical_policy: Option<String>,
    pub horizontal_stretch: Option<f64>,
    pub vertical_stretch: Option<f64>,
}

impl ConstantSizePolicy {
    /// Generates size policy from the given `binding_map`.
    ///
    /// The `cls` is supposed to be of `QSizePolicy` type.
    pub fn from_binding_map(
        cls: &Class,
        _node: Node,
        binding_map: &UiBindingMap,
        source: &str,
        diagnostics: &mut Diagnostics,
    ) -> Option<Self> {
        let mut policy = ConstantSizePolicy::default();
        let mut ok = true;
        for (&name, value) in binding_map {
            match name {
                "horizontalPolicy" => {
                    policy.horizontal_policy = extract_size_policy(cls, value, source, diagnostics);
                }
                "verticalPolicy" => {
                    policy.vertical_policy = extract_size_policy(cls, value, source, diagnostics);
                }
                "horizontalStretch" => {
                    policy.horizontal_stretch =
                        expr::evaluate_number(cls, value, source, diagnostics);
                }
                "verticalStretch" => {
                    policy.vertical_stretch =
                        expr::evaluate_number(cls, value, source, diagnostics);
                }
                _ => {
                    diagnostics.push(Diagnostic::error(
                        value.node().byte_range(),
                        format!(
                            "unknown property of class '{}': {}",
                            cls.qualified_name(),
                            name
                        ),
                    ));
                    ok = false;
                }
            }
        }

        if ok {
            Some(policy)
        } else {
            None
        }
    }

    /// Serializes this to UI XML.
    pub fn serialize_to_xml<W>(&self, writer: &mut XmlWriter<W>) -> XmlResult<()>
    where
        W: io::Write,
    {
        let mut tag = BytesStart::borrowed_name(b"sizepolicy");
        if let Some(s) = &self.horizontal_policy {
            tag.push_attribute(("hsizetype", s.as_ref()));
        }
        if let Some(s) = &self.vertical_policy {
            tag.push_attribute(("vsizetype", s.as_ref()));
        }
        writer.write_event(Event::Start(tag.to_borrowed()))?;

        if let Some(d) = self.horizontal_stretch {
            xmlutil::write_tagged_str(writer, "horstretch", d.to_string())?;
        }
        if let Some(d) = self.vertical_stretch {
            xmlutil::write_tagged_str(writer, "verstretch", d.to_string())?;
        }

        writer.write_event(Event::End(tag.to_end()))?;
        Ok(())
    }
}

fn extract_size_policy(
    cls: &Class,
    value: &UiBindingValue,
    source: &str,
    diagnostics: &mut Diagnostics,
) -> Option<String> {
    if let Some(ty) = cls.resolve_type_scoped("QSizePolicy::Policy") {
        expr::format_enum_expression(cls, &ty, value, source, diagnostics)
            .and_then(|s| s.strip_prefix("QSizePolicy::").map(|t| t.to_owned()))
    } else {
        diagnostics.push(Diagnostic::error(
            value.node().byte_range(),
            "QSizePolicy::Policy cannot be resolved from type map",
        ));
        None
    }
}
