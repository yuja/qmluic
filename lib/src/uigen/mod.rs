//! Qt user interface XML (.ui) generator.

use crate::diagnostic::{Diagnostic, Diagnostics};
use crate::qmlast::{UiImportSource, UiProgram};
use crate::qmldir;
use crate::qmldoc::UiDocument;
use crate::typemap::{Class, Module, ModuleData, ModuleId, Type, TypeMap, TypeSpace};
use thiserror::Error;

mod expr;
mod gadget;
mod layout;
mod object;
mod property;
mod xmlutil;

pub use self::expr::*; // re-export
pub use self::gadget::*; // re-export
pub use self::layout::*; // re-export
pub use self::object::*; // re-export

pub type XmlError = quick_xml::Error;
pub type XmlResult<T> = quick_xml::Result<T>;
pub type XmlWriter<W> = quick_xml::Writer<W>;

/// Builds `UiForm` from the given `doc`.
pub fn build(
    base_ctx: &BuildContext,
    doc: &UiDocument,
    diagnostics: &mut Diagnostics,
) -> Option<UiForm> {
    let program = diagnostics.consume_err(UiProgram::from_node(doc.root_node(), doc.source()))?;
    // TODO: do not build unnamed module space if directory module available
    let module_data = make_doc_module_data(doc, &program, base_ctx.type_map, diagnostics);
    let type_space = get_doc_type(base_ctx.type_map, doc)
        .unwrap_or_else(|| Type::Module(Module::new(&module_data, base_ctx.type_map)));
    let ctx = BuildDocContext::new(doc, type_space, base_ctx);
    let (obj, cls) =
        object::resolve_object_definition(&ctx, program.root_object_node(), diagnostics)?;
    let root_object =
        UiObject::from_object_definition(&ctx, &cls, &obj, ContainerKind::Any, diagnostics)?;
    Some(UiForm {
        class: doc.type_name().map(|s| s.to_owned()),
        root_object,
    })
}

fn get_doc_type<'a>(type_map: &'a TypeMap, doc: &UiDocument) -> Option<Type<'a>> {
    if let Some(doc_path) = doc.path() {
        let base_dir =
            qmldir::normalize_path(doc_path.parent().expect("doc path should point to file"));
        type_map
            .get_module(ModuleId::Directory(base_dir.into()))
            .and_then(|m| {
                m.get_type(
                    doc.type_name()
                        .expect("doc loaded from file should have type"),
                )
            })
    } else {
        None
    }
}

fn make_doc_module_data(
    doc: &UiDocument,
    program: &UiProgram,
    type_map: &TypeMap,
    diagnostics: &mut Diagnostics,
) -> ModuleData {
    let mut module_data = ModuleData::with_builtins();
    for imp in program.imports() {
        if imp.alias().is_some() {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "aliased import is not supported",
            ));
            continue;
        }
        // TODO: warn that version field is ignored
        // TODO: anchor diagnostic message onto imp.source() node
        let id = match imp.source() {
            UiImportSource::Identifier(x) => {
                ModuleId::Named(x.to_string(doc.source()).into_owned().into())
            }
            UiImportSource::String(x) => {
                if let Some(p) = doc.path().and_then(|p| p.parent()) {
                    ModuleId::Directory(qmldir::normalize_path(p.join(&x)).into())
                } else {
                    diagnostics.push(Diagnostic::error(
                        imp.node().byte_range(),
                        "cannot resolve directory path against inline QML document",
                    ));
                    continue;
                }
            }
        };
        if type_map.contains_module(&id) {
            module_data.import_module(id);
        } else {
            diagnostics.push(Diagnostic::error(
                imp.node().byte_range(),
                "module not found",
            ));
        }
    }
    module_data
}

/// Resources passed around the UI object constructors.
#[derive(Clone, Debug)]
pub struct BuildContext<'a> {
    type_map: &'a TypeMap,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
}

#[derive(Clone, Debug)]
struct BuildDocContext<'a, 's> {
    source: &'s str,
    action_class: Class<'a>,
    form_layout_class: Class<'a>,
    grid_layout_class: Class<'a>,
    hbox_layout_class: Class<'a>,
    layout_class: Class<'a>,
    layout_attached_class: Class<'a>,
    push_button_class: Class<'a>,
    spacer_item_class: Class<'a>,
    tab_widget_class: Class<'a>,
    tab_widget_attached_class: Class<'a>,
    vbox_layout_class: Class<'a>,
    widget_class: Class<'a>,
    type_space: Type<'a>,
}

impl<'a> BuildContext<'a> {
    /// Sets up context with the given `type_map`.
    pub fn prepare(type_map: &'a TypeMap) -> Result<Self, BuildContextError> {
        const MODULE_NAME: &str = "qmluic.QtWidgets";
        let module = type_map
            .get_module(&ModuleId::Named(MODULE_NAME.into()))
            .ok_or(BuildContextError::ModuleNotFound(MODULE_NAME))?;
        let get_class = |name| {
            if let Some(Type::Class(cls)) = module.get_type(name) {
                Ok(cls)
            } else {
                Err(BuildContextError::ClassNotFound(name))
            }
        };
        Ok(BuildContext {
            type_map,
            action_class: get_class("QAction")?,
            form_layout_class: get_class("QFormLayout")?,
            grid_layout_class: get_class("QGridLayout")?,
            hbox_layout_class: get_class("QHBoxLayout")?,
            layout_class: get_class("QLayout")?,
            layout_attached_class: get_class("QLayoutAttached")?,
            push_button_class: get_class("QPushButton")?,
            spacer_item_class: get_class("QSpacerItem")?,
            tab_widget_class: get_class("QTabWidget")?,
            tab_widget_attached_class: get_class("QTabWidgetAttached")?,
            vbox_layout_class: get_class("QVBoxLayout")?,
            widget_class: get_class("QWidget")?,
        })
    }
}

impl<'a, 's> BuildDocContext<'a, 's> {
    fn new(doc: &'s UiDocument, type_space: Type<'a>, base_ctx: &BuildContext<'a>) -> Self {
        BuildDocContext {
            source: doc.source(),
            action_class: base_ctx.action_class.clone(),
            form_layout_class: base_ctx.form_layout_class.clone(),
            grid_layout_class: base_ctx.grid_layout_class.clone(),
            hbox_layout_class: base_ctx.hbox_layout_class.clone(),
            layout_class: base_ctx.layout_class.clone(),
            layout_attached_class: base_ctx.layout_attached_class.clone(),
            push_button_class: base_ctx.push_button_class.clone(),
            spacer_item_class: base_ctx.spacer_item_class.clone(),
            tab_widget_class: base_ctx.tab_widget_class.clone(),
            tab_widget_attached_class: base_ctx.tab_widget_attached_class.clone(),
            vbox_layout_class: base_ctx.vbox_layout_class.clone(),
            widget_class: base_ctx.widget_class.clone(),
            type_space,
        }
    }
}

/// Error occurred while setting up [`BuildContext`].
#[derive(Clone, Debug, Error)]
pub enum BuildContextError {
    #[error("required class not found: {0}")]
    ClassNotFound(&'static str),
    #[error("required module not found: {0}")]
    ModuleNotFound(&'static str),
}
