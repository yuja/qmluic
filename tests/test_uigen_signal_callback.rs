use qmluic::uigen::DynamicBindingHandling;

pub mod common;

#[test]
fn test_unknown_signal() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QWidget {
        onWhatever: 0
    }
    "###).unwrap_err(), @r###"
    error: unknown signal of class 'QWidget': whatever
      ┌─ <unknown>:3:5
      │
    3 │     onWhatever: 0
      │     ^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_invalid_binding_map_callback() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QDialog {
        onAccepted.whatever: 0
    }
    "###).unwrap_err(), @r###"
    error: signal callback cannot be a map
      ┌─ <unknown>:3:16
      │
    3 │     onAccepted.whatever: 0
      │                ^^^^^^^^
    "###);
}

#[test]
fn test_callback_without_dynamic_binding() {
    insta::assert_snapshot!(common::translate_str(r###"
    import qmluic.QtWidgets
    QDialog {
        onAccepted: 0
    }
    "###).unwrap_err(), @r###"
    error: signal callback cannot be translated without dynamic binding
      ┌─ <unknown>:3:5
      │
    3 │     onAccepted: 0
      │     ^^^^^^^^^^^^^
    "###);
}

#[test]
fn test_root_connection() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QDialog {
            onAccepted: 0
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}

#[test]
fn test_non_root_connection() {
    let doc = common::parse_doc(
        r###"
        import qmluic.QtWidgets
        QDialog {
            QDialogButtonBox {
                onAccepted: 0
            }
        }
        "###,
    );
    let (_, ui_support_h) = common::translate_doc(&doc, DynamicBindingHandling::Generate).unwrap();
    insta::assert_snapshot!(ui_support_h);
}
