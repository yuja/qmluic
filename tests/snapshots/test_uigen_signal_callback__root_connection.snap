---
source: tests/test_uigen_signal_callback.rs
expression: ui_support_h
---
#pragma once
#include "ui_mytype.h"

namespace UiSupport {
class MyType
{
public:
    MyType(QDialog *root, Ui::MyType *ui): root_(root), ui_(ui) {}

    void setup()
    {
        this->setupDialogAccepted();
    }

private:
    enum class BindingIndex : unsigned {
    };

    void setupDialogAccepted()
    {
        QObject::connect(this->root_, &QDialog::accepted, root_, [this]() { this->onDialogAccepted(); });
    }

    void onDialogAccepted()
    {
    b0:
        static_cast<void>(0);
        return;
    }

    QDialog *const root_;
    Ui::MyType *const ui_;

};
} // namespace UiSupport

